crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,516 @@
+        // called to write the remaining data to the newly provided buffers. If there is no
+        // remaining data waiting to be sent it can return 0.
+        if input_buffer.is_empty() {
+            return Ok(0);
+        }
+
+        // Break up the data from the "input_buffer" into chunks that fit into the buffers of the
+        // receive queue descriptor chain and write it to guest memory.
+        while let Some(desc) = desc_chain.next() {
+            if !desc.is_write_only() {
+                return Err(Error::UnexpectedReadOnlyDescriptor);
+            }
+            let len_write = cmp::min(desc.len() as usize, input_buffer.len());
+            let buffer_slice = input_buffer.drain(..len_write).collect::<Vec<u8>>();
+
+            desc_chain
+                .memory()
+                .write_slice(&buffer_slice, desc.addr())
+                .map_err(Error::WriteToGuestFailed)?;
+
+            // Should not overflow since "input_buffer" can hold a maximum of usize::MAX elements
+            sent_bytes += len_write;
+
+            if input_buffer.is_empty() {
+                break;
+            }
+        }
+
+        Ok(sent_bytes)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use virtio_queue::defs::VIRTQ_DESC_F_WRITE;
+    use virtio_queue::mock::MockSplitQueue;
+    use virtio_queue::Descriptor;
+    use vm_memory::{GuestAddress, GuestMemoryMmap};
+
+    impl PartialEq for Error {
+        fn eq(&self, other: &Self) -> bool {
+            use self::Error::*;
+            match (self, other) {
+                (WriteToGuestFailed(ref e), WriteToGuestFailed(ref other_e)) => {
+                    format!("{}", e).eq(&format!("{}", other_e))
+                }
+                (WriteToOutputFailed(ref e), WriteToOutputFailed(ref other_e)) => {
+                    format!("{}", e).eq(&format!("{}", other_e))
+                }
+                (ChainLengthOverflow, ChainLengthOverflow) => true,
+                (BufferCapacityExceeded, BufferCapacityExceeded) => true,
+                (UnexpectedReadOnlyDescriptor, UnexpectedReadOnlyDescriptor) => true,
+                (UnexpectedWriteOnlyDescriptor, UnexpectedWriteOnlyDescriptor) => true,
+                (InvalidBufferCapacity, InvalidBufferCapacity) => true,
+                _ => false,
+            }
+        }
+    }
+
+    const INPUT_VALUE: u8 = 42;
+    const INPUT_SIZE: usize = 256;
+
+    #[test]
+    fn test_input_buffer_operations() {
+        assert_eq!(
+            Console::new_with_capacity(0, vec![0; DEFAULT_CAPACITY]).unwrap_err(),
+            Error::InvalidBufferCapacity
+        );
+
+        assert_eq!(
+            Console::new_with_capacity(MAX_CAPACITY + 1, vec![0; DEFAULT_CAPACITY]).unwrap_err(),
+            Error::InvalidBufferCapacity
+        );
+
+        let console =
+            Console::new_with_capacity(DEFAULT_CAPACITY, vec![0; DEFAULT_CAPACITY]).unwrap();
+        assert!(!console
```

Comment:
Instead of `assert!(!....is_err())`s you can use `assert!(....is_ok())`.

Reply:
We can also just use `unwrap` instead of assert for added code readability:

```rust
let console =
            Console::new_with_capacity(DEFAULT_CAPACITY, vec![0; DEFAULT_CAPACITY]).unwrap();
console.enqueue_data(&mut vec![INPUT_VALUE; DEFAULT_CAPACITY]).unwrap();
....
```